contracts/src/PDPService.sol

```diff
@@ -303,7 +302,29 @@ contract PDPService {
     // by the epoch of the previous proof of possession.
     // Note that this method is not restricted to the proof set owner.
     function provePossession(uint256 setId, Proof[] calldata proofs) public {
-        // TODO: implement me
+        uint256 challengeEpoch = nextChallengeEpoch[setId];
+        require(block.number >= challengeEpoch, "premature proof");
+
+        // TODO: fetch proper seed from chain randomness
+        uint256 seed = challengeEpoch;
+        uint256 leafCount = getProofSetLeafCount(setId);
+        uint256 sumTreeTop = 256 - BitOps.clz(nextRootId[setId]);
+
+        for (uint64 i = 0; i < proofs.length; i++) {
+            // Hash (SHA3) the seed,  proof set id, and proof index to create challenge.
+            bytes memory payload = abi.encodePacked(seed, setId, i);
+            uint256 challengeIdx = uint256(keccak256(payload)) % leafCount;
+
+            // Find the root that has this leaf, and the offset of the leaf within that root.
+            RootIdAndOffset memory root = findOneRootId(setId, challengeIdx, sumTreeTop);
+            bytes32 rootHash = Cids.digestFromCid(getRootCid(setId, root.rootId));
+            bool ok = MerkleVerify.verify(proofs[i].proof, rootHash, proofs[i].leaf, root.offset);
+            require(ok, "proof did not verify");
+        }
+
+        // Set the next challenge epoch.
+        nextChallengeEpoch[setId] = block.number + challengeFinality; 
+        // TODO: record the successful proof somewhere.
```

Comment:
You can remove this as the record keeper PR should reflect these requirements.  I hadn't planned on storing the actual proof but maybe we should at least add it into the recordkeeper call for the consumer to potentially keep around.

Reply:
I just meant record that the proof happened.